File: Python/decompress.py
def decompress(compressed):
    result, int_stack, str_stack, n = '', [], [], 0
    for ch in compressed:
        if ch.isdigit():
            n = n * 10 + int(ch)
        elif ch == '[':
            int_stack.append(n)
            str_stack.append(result)
            result, n = '', 0
        elif ch == ']':
            n, s = int_stack.pop(), str_stack.pop()
            result = s + n * result
            n = 0
        else:
            result += ch
    return result


if __name__ == '__main__':
  print(decompress('3[abc]4[ab]c')) # -> abcabcabcababababc
  print(decompress('10[a]'))        # -> aaaaaaaaaa
  print(decompress('2[3[a]b]'))     # -> aaabaaab
  print(decompress('2[c3[a]b]'))    # -> caaabcaaab

File: Java/Decompress.java
import java.util.Stack;

public class Decompress {
  public static String decompress(String compressed) {
    Stack<Integer> intStack = new Stack<>();
    Stack<StringBuilder> strStack = new Stack<>();
    StringBuilder cur = new StringBuilder();
    int k = 0;
    for (int i = 0; i < compressed.length(); i++) {
        char ch = compressed.charAt(i);
        if (Character.isDigit(ch)) {
            k = k * 10 + ch - '0';
        } else if (ch == '[') {
            intStack.push(k);
            strStack.push(cur);
            cur = new StringBuilder();
            k = 0;
        } else if (ch == ']') {
            StringBuilder tmp = cur;
            cur = strStack.pop();
            for (k = intStack.pop(); k > 0; --k) cur.append(tmp);
        } else cur.append(ch);
    }
    return cur.toString();
  }
  
  public static void main(String[] args) {
    System.out.println(decompress("3[abc]4[ab]c")); // -> abcabcabcababababc
    System.out.println(decompress("10[a]"));        // -> aaaaaaaaaa
    System.out.println(decompress("2[3[a]b]"));     // -> aaabaaab
    System.out.println(decompress("2[c3[a]b]"));    // -> caaabcaaab
  }
}
